src/ir/module-splitting.cpp

```diff
@@ -0,0 +1,405 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ir/module-splitting.h"
+#include "ir/manipulation.h"
+#include "ir/module-utils.h"
+#include "ir/names.h"
+#include "pass.h"
+#include "wasm-builder.h"
+#include "wasm.h"
+
+namespace wasm {
+
+namespace ModuleSplitting {
+
+namespace {
+
+std::unique_ptr<Module> initializeSecondary(const Module& primary) {
+  // Create the secondary module and copy trivial properties.
+  auto secondary = std::make_unique<Module>();
+  secondary->features = primary.features;
+  secondary->hasFeaturesSection = primary.hasFeaturesSection;
+  return secondary;
+}
+
+void shareImportableItems(Module& primary,
+                          Module& secondary,
+                          Name importNamespace) {
+  // Map internal names to (one of) their corresponding export names.
+  std::unordered_map<Name, Name> exports;
+  for (auto& ex : primary.exports) {
+    if (ex->kind != ExternalKind::Function) {
+      exports[ex->value] = ex->name;
+    }
+  }
+
+  auto makeImportExport = [&](Importable& primaryItem,
+                              Importable& secondaryItem,
+                              Name exportName,
+                              ExternalKind kind) {
+    secondaryItem.name = primaryItem.name;
+    secondaryItem.hasExplicitName = primaryItem.hasExplicitName;
+    secondaryItem.module = importNamespace;
+    auto exportIt = exports.find(primaryItem.name);
+    if (exportIt != exports.end()) {
+      secondaryItem.base = exportIt->second;
+    } else {
+      exportName = Names::getValidExportName(primary, exportName);
+      primary.addExport(new Export{exportName, primaryItem.name, kind});
+      secondaryItem.base = exportName;
+    }
+  };
+
+  // TODO: Be more selective by only sharing global items that are actually used
+  // in the secondary module, just like we do for functions.
+
+  auto shareMemory = [&](Memory* memory) {
+    secondary.memory.exists = true;
+    secondary.memory.initial = memory->initial;
+    secondary.memory.max = memory->max;
+    secondary.memory.shared = memory->shared;
+    secondary.memory.indexType = memory->indexType;
+    makeImportExport(*memory, secondary.memory, "memory", ExternalKind::Memory);
+  };
+  ModuleUtils::iterImportedMemories(primary, shareMemory);
+  ModuleUtils::iterDefinedMemories(primary, shareMemory);
+
+  auto shareTable = [&](Table* table) {
+    secondary.table.exists = true;
+    secondary.table.initial = table->initial;
+    secondary.table.max = table->max;
+    makeImportExport(*table, secondary.table, "table", ExternalKind::Table);
+  };
```

Comment:
I wouldn't call this deliberate, per se. I don't know where I picked up this coding style, but I like it because it limits the scope in which functions are available to the scope in which they are used. It also means you have just one place to look for variable declarations (previously in the function) rather than two (the function and the enclosing class).

I do think I could break up the function a bit, though. I was also considering using an RAII-style context manager to manage the new table slot allocations. I'll refactor this a bit and see if it seems any cleaner.

Reply:
You may have picked it up here; Binaryen has always used it extensively; and I suspect @kripken picked it up from JS.